### PR TITLE
Stats collector using context managers to hold the data

### PR DESCRIPTION
**Pull request by kkalinowski-reef**
- instead of lists, were using a class that behaves kinda like a list, but is also a context manager that measures time;
- total moved into a context manager as well, class for stats handles all the logic now;
- data fetching moved from a `for` to a generator `while`, to be able to enclose it in context manager easily;

Warning!
This code is still not pretty. But it's easier to count indentation than to remember to bump every value (one was missing).

### CODE REVIEW

File: b2sdk/transfer/inbound/downloader/parallel.py [hunk 1]
```diff
@@ -294,67 +292,74 @@ def download_first_part(
...
-            before_read = perf_counter_ns()
-            for to_write in response.iter_content(chunk_size=chunk_size):
-                stats_collector_read_append(perf_counter_ns() - before_read)
 
-                before_put = perf_counter_ns()
+    with stats_collector.total:
+
+        stats_collector_read = stats_collector.read
+        stats_collector_other = stats_collector.other
+        stats_collector_write = stats_collector.write
```
Comment (ppolewicz) [inline b2sdk/transfer/inbound/downloader/parallel.py:300]: cosmetic: I'd move that to right after `stats_collector` is initialized, breaking the principle where we should declare variables as later as possible - I wouldn't want to muddy the code which actually does things with taking off the `.` to improve performance. As it is now, you are measuring those lines as part of "total" which may not skew the results too much, but still, it's like a lock, don't keep it longer than you need to.

File: b2sdk/transfer/inbound/downloader/stats_collector.py [hunk 2]
```diff
@@ -9,42 +9,71 @@
...
+        time_diff = perf_counter_ns() - self.started_perf_timer
+        self.storage.append(time_diff)
+        self.started_perf_timer = None
+
+    def __getitem__(self, item_or_slice):
+        return self.storage[item_or_slice]
+
+    @property
+    def has_any_entry(self) -> bool:
+        return len(self.storage) > 0
```
Comment (ppolewicz) [inline b2sdk/transfer/inbound/downloader/stats_collector.py:45]:
```suggestion
        return self.storage
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -22,6 +22,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Fix nox's deprecated `session.install()` calls
 * Re-enable changelog validation in CI
 
+### Changed
```
Comment (ppolewicz) [inline CHANGELOG.md:25]: it's not a public interface the library users care about, therefore please put it in the `### Infrastructure` section

File: b2sdk/transfer/inbound/downloader/stats_collector.py (hunk 2 above)
Comment (kkalinowski-reef) [reply b2sdk/transfer/inbound/downloader/stats_collector.py:45]: Ok, we're going for the faster solution. Fair enough.

File: b2sdk/transfer/inbound/downloader/parallel.py (hunk 1 above)
Comment (kkalinowski-reef) [reply b2sdk/transfer/inbound/downloader/parallel.py]: Done.

File: b2sdk/transfer/inbound/downloader/stats_collector.py (hunk 2 above)
Comment (kkalinowski-reef) [reply b2sdk/transfer/inbound/downloader/stats_collector.py]: Done.

File: CHANGELOG.md (hunk 3 above)
Comment (kkalinowski-reef) [reply CHANGELOG.md]: Done.

**Pull request closed without merging**